llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -4018,5 +4018,30 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
   if (CondVal->getType() == SI.getType() && isKnownInversion(FalseVal, TrueVal))
     return BinaryOperator::CreateXor(CondVal, FalseVal);
 
+  if (SelType->isIntOrIntVectorTy()) {
+    // Try to simplify select arms based on KnownBits implied by the condition.
+    CondContext CC(CondVal);
+    findValuesAffectedByCondition(CondVal, /*IsAssume=*/false, [&](Value *V) {
+      CC.AffectedValues.insert(V);
+    });
```

Comment:
Maybe aggressive inst combine is the right place for this?
Maybe we could also track `select` instructions in DomConditionCache (although that would req some more complex logic to keep it updated as we created/delect select instructions).

Also, think for select you can at least limit `V` to values that are actually used by select arms.
Maybe create a set of uses up to Depth = 6 for each of the arms and only add affected values if they also hit that set? Otherwise can't really imagine this helping.

You might also try early out of `!isa<Constant>(TrueVal) && !isa<Constant>(FalseVal)` to avoid unnecessary setup.

Finally, maybe try simplifying both `True` and `False` arm in one go instead of essentially doubling the number of trips.

Reply:
Actually, think you really should only add `V` if its `TrueArm`/`FalseArm`.

Reply:
As mentioned in the PR description, most of the overhead comes from simply adding a member to SimplifyQuery rather than anything happening in InstCombine. These are the numbers just for adding a dummy member: https://llvm-compile-time-tracker.com/compare.php?from=3ca17443ef4af21bdb1f3b4fbcfff672cbc6176c&to=e7a64d837a061d6afeac9c0f06c4827998d43561&stat=instructions:u

As such, I don't think we'll get any substantial improvement out of changing how exactly the KnownBits calculation is done.

> Maybe aggressive inst combine is the right place for this?

This would prevent extending this to use SimplifyDemandedBits, which enabled simplifying the expression without folding to constants (https://github.com/nikic/llvm-project/commit/1b8edbd1178b08fc544c6af4e1826a6b51262485). This will allow us to subsume special cases like https://github.com/llvm/llvm-project/pull/92658.

> Maybe we could also track `select` instructions in DomConditionCache (although that would req some more complex logic to keep it updated as we created/delect select instructions).

I think managing invalidation for this would be quite tricky, and this approach would be inherently limited to one-use chains to the select only.

> Actually, think you really should only add `V` if its `TrueArm`/`FalseArm`.

Why? V may be used by TrueArm/FalseArm recursively.

Reply:
> As mentioned in the PR description, most of the overhead comes from simply adding a member to SimplifyQuery rather than anything happening in InstCombine. These are the numbers just for adding a dummy member: https://llvm-compile-time-tracker.com/compare.php?from=3ca17443ef4af21bdb1f3b4fbcfff672cbc6176c&to=e7a64d837a061d6afeac9c0f06c4827998d43561&stat=instructions:u
> 
> As such, I don't think we'll get any substantial improvement out of changing how exactly the KnownBits calculation is done.
> 
I see, although a bit more ugly, what about adding an additional optional argument to `computeKnownBits`/`computeKnownBitsFromContext`?

Reply:
I didn't get what you refer to with "saving an iteration of InstCombine" -- how can we save an InstCombine iteration?

For the record, this is the impact of the KnownBits calculation (as opposed to the extra member and the affected value calculation): http://llvm-compile-time-tracker.com/compare.php?from=c9edd08d49137e47fe3acda3ea5b2bae563cb3ac&to=db730283063dff0dc72338413f346bd0174dafdd&stat=instructions:u

Reply:
> I didn't get what you refer to with "saving an iteration of InstCombine" -- how can we save an InstCombine iteration?
> 

I mean do the full transform in a single shot as opposed to going through the work list twice.

Iteration was wrong word to us there given it's other meaning.

Reply:
> I mean do the full transform in a single shot as opposed to going through the work list twice.

Do you mean the case where it's possible to simplify *both* select operands? If so, this should be very rare, and as such not worth optimizing.